### PR TITLE
pkg/bpf/ctmap: refactor to utilize pkg/bpf/map

### PR DESCRIPTION
**Pull request by ianvernon**
Refactor pkg/bpf/ctmap to utilize pkg/bpf/map implementation.
Add GetNextKey wrapper in pkg/bpf/map.go.

Signed-off by: Ian Vernon <ian@covalent.io>

Fixes: #357

### CODE REVIEW

**Comment by ianvernon** [comment]
I would like to only have one map refactoring per issue - 357 says it is for all maps, but would be better / more modular if we have multiple tickets.

File: cilium/cmd/bpf_ct_list.go [hunk 1]
```diff
@@ -17,9 +17,9 @@ package cmd
 import (
 	"fmt"
 
-	"github.com/cilium/cilium/pkg/bpf"
 	"github.com/cilium/cilium/pkg/maps/ctmap"
 
+	"github.com/cilium/cilium/pkg/bpf"
```
Comment (aanm) [inline cilium/cmd/bpf_ct_list.go]: This should be grouped with the remaining `github.com/cilium/cilium` imports

File: cilium/cmd/bpf_ct_list.go [hunk 2]
```diff
@@ -30,11 +30,11 @@ var bpfCtListCmd = &cobra.Command{
 	PreRun: requireEndpointIDorGlobal,
 	Run: func(cmd *cobra.Command, args []string) {
 		if args[0] == "global" {
-			dumpCtProto(ctmap.MapName6+args[0], ctmap.CtTypeIPv6Global)
-			dumpCtProto(ctmap.MapName4+args[0], ctmap.CtTypeIPv4Global)
+			dumpCtProto(ctmap.MapName6Global, "")
```
Comment (aanm) [inline cilium/cmd/bpf_ct_list.go]: Why the `dumpCtProto` from `else` take the `args[0]` but this one doesn't take the `"global"` string?

File: cilium/cmd/bpf_ct_list.go [hunk 3]
```diff
@@ -43,19 +43,18 @@ func init() {
...
 	if err != nil {
-		Fatalf("Unable to open %s: %s\n", file, err)
+		Fatalf("Unable to open map %s: %s", file, err)
 	}
-
-	m := ctmap.CtMap{Fd: fd, Type: ctType}
-	out, err := m.Dump()
+	out, err := ctmap.ToString(m, mapType)
 	if err != nil {
 		Fatalf("Error while dumping BPF Map: %s\n", err)
```
Comment (aanm) [inline cilium/cmd/bpf_ct_list.go]: You can remove the `\n` here as well

File: daemon/ct.go [hunk 4]
```diff
@@ -30,44 +29,45 @@ const (
...
+		} else {
+			mapType = ctmap.MapName4Global
 		}
-	}
-	fd, err := bpf.ObjGet(file)
-	if err != nil {
-		log.Warningf("Unable to open CT map %s: %s\n", file, err)
-		e.LogStatus(endpoint.BPF, endpoint.Warning, fmt.Sprintf("Unable to open CT map %s: %s", file, err))
-		return
+		file = bpf.MapPath(mapType)
```
Comment (aanm) [inline daemon/ct.go]: Won't this overwrite the variable written on line 49?

File: pkg/bpf/map.go [hunk 5]
```diff
@@ -405,3 +400,30 @@ func (m *Map) DeleteAll() error {
 
 	return nil
 }
+
+// openIfNotOpened opens Map m if it isn't opened.
+func (m *Map) openIfNotOpened() error {
+	if m.fd == 0 {
+		if err := m.Open(); err != nil {
```
Comment (aanm) [inline pkg/bpf/map.go]: you can replace this with by simply having `return m.Open()`

File: pkg/bpf/map.go [hunk 6]
```diff
@@ -405,3 +400,30 @@ func (m *Map) DeleteAll() error {
...
+			return err
+		}
+	}
+
+	return nil
+}
+
+//GetNextKey returns the next key in the Map after key.
+func (m *Map) GetNextKey(key MapKey, nextKey MapKey) error {
+	if m.fd == 0 {
```
Comment (aanm) [inline pkg/bpf/map.go]: Why not use the `m.openIfNotOpened()`?

File: pkg/maps/ctmap/ctmap.go [hunk 7]
```diff
@@ -54,154 +48,32 @@ const (
...
-type CtKey6Global struct {
-	daddr   types.IPv6
-	saddr   types.IPv6
-	sport   uint16
-	dport   uint16
-	nexthdr u8proto.U8proto
-	flags   uint8
-}
+// ServiceKey is the interface describing protocol independent key for services map.
+type ServiceKey interface {
```
Comment (aanm) [inline pkg/maps/ctmap/ctmap.go]: Why the name `ServiceKey`? Is this related to the notion of virtual IPs for the LB that we have?

File: pkg/maps/ctmap/ctmap.go [hunk 8]
```diff
@@ -17,19 +17,13 @@ package ctmap
 import (
 	"bytes"
 	"fmt"
-	"unsafe"
-
 	"github.com/cilium/cilium/common"
-	"github.com/cilium/cilium/common/types"
 	"github.com/cilium/cilium/pkg/bpf"
-	"github.com/cilium/cilium/pkg/u8proto"
+	"github.com/op/go-logging"
```
Comment (aanm) [inline pkg/maps/ctmap/ctmap.go]: should be:
```
	"bytes"
	"fmt"
	"unsafe"

	"github.com/cilium/cilium/common"
	"github.com/cilium/cilium/pkg/bpf"

	"github.com/op/go-logging"
```

File: pkg/maps/ctmap/ctmap.go [hunk 9]
```diff
@@ -54,154 +48,32 @@ const (
 	CtTypeIPv4Global
```
Comment (aanm) [inline pkg/maps/ctmap/ctmap.go]: All these 4 const are unused

File: pkg/maps/ctmap/ctmap.go [hunk 10]
```diff
@@ -250,156 +130,155 @@ func (m *CtMap) Dump() (string, error) {
...
 			key = nextKey
 		}
 	}
-
 	return entries, nil
 }
 
-func (m *CtMap) doGc(interval uint16, key unsafe.Pointer, nextKey unsafe.Pointer, deleted *int) bool {
-	var entry CtEntry
+// doGC removes key and its corresponding value from Map m if key has a lifetime less than interval. It sets nextKey's address to be the key after key in map m.
```
Comment (aanm) [inline pkg/maps/ctmap/ctmap.go]: Small nit, can you write this in 80 columns?

File: pkg/maps/ctmap/ctmap.go [hunk 11]
```diff
@@ -250,156 +130,155 @@ func (m *CtMap) Dump() (string, error) {
...
+// Returns false if there is no other key after key in Map m, and increments deleted if an item is removed from m.
+func doGc(m *bpf.Map, interval uint16, key bpf.MapKey, nextKey bpf.MapKey, deleted *int) bool {
+	err := m.GetNextKey(key, nextKey)
 
-	err := bpf.GetNextKey(m.Fd, key, nextKey)
+	// GetNextKey errors out if we have reached the last entry in the map. This signifies that the entire map has been traversed and that garbage collection is done.
```
Comment (aanm) [inline pkg/maps/ctmap/ctmap.go]: Small nit, can you write this in 80 columns?

File: pkg/maps/ctmap/ctmap.go [hunk 12]
```diff
@@ -250,156 +130,155 @@ func (m *CtMap) Dump() (string, error) {
...
 		entry.lifetime -= interval
-		bpf.UpdateElement(m.Fd, nextKey, unsafe.Pointer(&entry), 0)
+		m.Update(nextKey, entry)
 	}
 
 	return true
 }
 
-func (m *CtMap) GC(interval uint16) int {
+// GC runs garbage collection for map m with name mapName with interval interval. It returns how many items were deleted from m.
```
Comment (aanm) [inline pkg/maps/ctmap/ctmap.go]: Small nit, can you write this in 80 columns?

File: pkg/maps/ctmap/ctmap.go [hunk 13]
```diff
@@ -250,156 +130,155 @@ func (m *CtMap) Dump() (string, error) {
...
+
 	if err != nil {
+		log.Errorf("error during map Lookup: %s", err)
 		return false
 	}
 
+	entry := nextEntry.(*CtEntry)
 	if entry.lifetime <= interval {
-		bpf.DeleteElement(m.Fd, nextKey)
+		m.Delete(nextKey)
```
Comment (aanm) [inline pkg/maps/ctmap/ctmap.go]: `err = m.Delete(nextKey)`

File: pkg/maps/ctmap/ctmap.go [hunk 14]
```diff
@@ -250,156 +130,155 @@ func (m *CtMap) Dump() (string, error) {
...
+		if err != nil {
+			log.Debugf("error during Delete: %s", err)
+
+		}
 		(*deleted)++
 	} else {
 		entry.lifetime -= interval
-		bpf.UpdateElement(m.Fd, nextKey, unsafe.Pointer(&entry), 0)
+		m.Update(nextKey, entry)
```
Comment (aanm) [inline pkg/maps/ctmap/ctmap.go]: No debug message here? :-)
```
if err != nil {
			log.Debugf("error during Update: %s", err)
		}
```

File: pkg/maps/ctmap/ipv4.go [hunk 15]
```diff
@@ -0,0 +1,160 @@
+package ctmap
+
+import (
+	"bytes"
```
Comment (aanm) [inline pkg/maps/ctmap/ipv4.go:18]:
```
	"bytes"
	"fmt"
	"net"
	"unsafe"

	"github.com/cilium/cilium/common"
	"github.com/cilium/cilium/common/types"
	"github.com/cilium/cilium/pkg/bpf"
	"github.com/cilium/cilium/pkg/u8proto"
```

File: pkg/maps/ctmap/ipv4.go [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+	"github.com/cilium/cilium/common/types"
+	"github.com/cilium/cilium/pkg/bpf"
+	"github.com/cilium/cilium/pkg/u8proto"
+	"net"
+	"unsafe"
+)
+
+var (
+	// Service4Map is the map representing the bpf local conntrack map.
+	Service4Map = bpf.NewMap(MapName4+"_services",
```
Comment (aanm) [inline pkg/maps/ctmap/ipv4.go]: @tgraf I'm lost, do we have a services map for CT?

File: pkg/maps/ctmap/ipv4.go [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+		bpf.MapTypeHash,
+		int(unsafe.Sizeof(CtKey4{})),
+		int(unsafe.Sizeof(CtEntry{})),
+		MapNumEntriesLocal)
+
+	// Service4GlobalMap is the map representing the bpf global conntrack map.
+	Service4GlobalMap = bpf.NewMap(MapName4Global+"services",
```
Comment (aanm) [inline pkg/maps/ctmap/ipv4.go]: The name of `Service4Map` is `MapName4+"_services"` this one doesn't have an underscore, on purpose?

File: pkg/maps/ctmap/ipv4.go [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+type CtKey4 struct {
+	addr    types.IPv4
+	sport   uint16
+	dport   uint16
+	nexthdr u8proto.U8proto
+	flags   uint8
+}
+
+// NewCtKey4 creates a CtKey4 with the provided ip, source port, destination port, next header, and flags.
+func NewCtKey4(addr net.IP, sport uint16, dport uint16, nexthdr u8proto.U8proto, flags uint8) *CtKey4 {
```
Comment (aanm) [inline pkg/maps/ctmap/ipv4.go]: Seems this function is unused

File: pkg/maps/ctmap/ipv4.go [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+	n.sport = common.Swab16(n.sport)
+	n.dport = common.Swab16(n.dport)
+	return &n
+}
+
+func (k *CtKey4) String() string {
+	return fmt.Sprintf("%s:%d, %d, %d, %d", k.addr, k.sport, k.dport, k.nexthdr, k.flags)
+}
+
+// Dump writes the contents of key to buffer and returns true if the value for next header in the key is nonzero.
```
Comment (aanm) [inline pkg/maps/ctmap/ipv4.go]: 80 columns

File: pkg/maps/ctmap/ipv4.go [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+}
+
+// GetKeyPtr returns the unsafe.Pointer for k.
+func (k *CtKey4) GetKeyPtr() unsafe.Pointer { return unsafe.Pointer(k) }
+
+//NewValue creates a new bpf.MapValue.
+func (k *CtKey4) NewValue() bpf.MapValue { return &CtEntry{} }
+
+// Map returns the BPF map for the local IPv4 conntrack map.
+func (k CtKey4) Map() *bpf.Map { return Service4Map }
```
Comment (aanm) [inline pkg/maps/ctmap/ipv4.go]: @tgraf I'm not entirely sure that we should be returning something like this, if this is a local map, why are they being pointed to the same name?

File: pkg/maps/ctmap/ipv6.go [hunk 21]
```diff
@@ -0,0 +1,160 @@
...
+	n.dport = common.Swab16(n.dport)
+	return &n
+}
+
+func (k *CtKey6) String() string {
+	return fmt.Sprintf("%s:%d, %d, %d, %d", k.addr, k.sport, k.dport, k.nexthdr, k.flags)
+}
+
+// Dump writes the contents of key to buffer and returns true if the value for next header in the key is nonzero.
+func (key CtKey6) Dump(buffer *bytes.Buffer) bool {
```
[the remark this answers is not in the document]
Comment (aanm) [reply pkg/maps/ctmap/ipv6.go]: @ianvernon He wants you to have `(k CtKey6)` instead of `(key CtKey6)`

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/maps/ctmap/ipv4.go (hunk 16 above)
Comment (tgraf) [reply pkg/maps/ctmap/ipv4.go]: Hehe, no. I think this is just an oversight and can be removed.

File: pkg/maps/ctmap/ctmap.go (hunk 7 above)
Comment (tgraf) [reply pkg/maps/ctmap/ctmap.go]: This should be renamed to `CtKey` which is the interface for the common functions of `CtKey4` and `CtKey6`

File: pkg/maps/ctmap/ipv4.go (hunk 17 above)
Comment (tgraf) [reply pkg/maps/ctmap/ipv4.go]: I think that's jus a leftover from basing this of lbmap and can be removed.

File: pkg/maps/ctmap/ipv4.go (hunk 20 above)
Comment (tgraf) [reply pkg/maps/ctmap/ipv4.go]: lbmap uses `Map()` to hide name for IPv4 and IPv6 specific map names for services map. The difference is that lbmaps are always global, just per address family whereas ctmap is per endpoint unless global.

I think all of the `Map()` usage can be removed and what remains is the address family specific `Convert()`.

**Comment by tgraf** [comment]
Great work! This exactly the direction that I meant when I opened the issue. Agreed with handling one map at a time. Let's iron out the details here and use it as the new prime example of how to abstract BPF maps.

File: cilium/cmd/bpf_ct_list.go (hunk 2 above)
Comment (ianvernon) [reply cilium/cmd/bpf_ct_list.go]: cramp.MapName6Global is assigned the following value: MapName6Global = MapName6 + "global" so there is no need to use args[0] here.

File: daemon/ct.go (hunk 4 above)
Comment (ianvernon) [reply daemon/ct.go]: Line 49 is no longer needed - I'll remove it in the next upload.

File: pkg/bpf/map.go (hunk 5 above)
Comment (ianvernon) [reply pkg/bpf/map.go]: You're right - I don't think we really need this function anymore. I'll refactor accordingly.

File: pkg/bpf/map.go (hunk 6 above)
Comment (ianvernon) [reply pkg/bpf/map.go]: I'm replacing calls to openIfNotOpened with ```if err := m.Open(); err != nil {
			return err
		}``` in the next upload

File: pkg/maps/ctmap/ipv4.go (hunk 16 above)
Comment (ianvernon) [reply pkg/maps/ctmap/ipv4.go]: OK, I'll remove these maps.

File: pkg/maps/ctmap/ipv4.go (hunk 20 above)
Comment (ianvernon) [reply pkg/maps/ctmap/ipv4.go]: Ah, that makes sense. Thanks for clarifying; I will remove this.

File: cilium/cmd/bpf_ct_list.go (hunk 2 above)
Comment (tgraf) [reply cilium/cmd/bpf_ct_list.go]: I'm fine with this. Less dynamic calculation.

File: cilium/cmd/bpf_ct_list.go [hunk 22]
```diff
@@ -43,19 +43,18 @@ func init() {
...
 
-	fd, err := bpf.ObjGet(file)
 	if err != nil {
-		Fatalf("Unable to open %s: %s\n", file, err)
+		Fatalf("Unable to open map %s: %s", file, err)
 	}
-
-	m := ctmap.CtMap{Fd: fd, Type: ctType}
-	out, err := m.Dump()
+	out, err := ctmap.ToString(m, mapType)
```
Comment (tgraf) [inline cilium/cmd/bpf_ct_list.go:62]: 👍

File: pkg/maps/ctmap/ctmap.go [hunk 23]
```diff
@@ -250,156 +125,164 @@ func (m *CtMap) Dump() (string, error) {
 	return buffer.String(), nil
 }
 
-func (m *CtMap) DumpToSlice() ([]CtEntryDump, error) {
-	var entry CtEntry
+// DumpToSlice iterates through map m and returns a slice mapping each key to
+// its value in m.
+func DumpToSlice(m *bpf.Map, mapType string) ([]CtEntryDump, error) {
```
Comment (tgraf) [inline pkg/maps/ctmap/ctmap.go]: Can this be made private?

File: pkg/maps/ctmap/ctmap.go [hunk 24]
```diff
@@ -250,156 +125,164 @@ func (m *CtMap) Dump() (string, error) {
...
+		log.Errorf("error during map Lookup: %s", err)
 		return false
 	}
 
+	entry := nextEntry.(*CtEntry)
 	if entry.lifetime <= interval {
-		bpf.DeleteElement(m.Fd, nextKey)
+		err := m.Delete(nextKey)
+		if err != nil {
+			log.Debugf("error during Delete: %s", err)
```
Comment (tgraf) [inline pkg/maps/ctmap/ctmap.go:238]: 👍

**Review by tgraf — CHANGES_REQUESTED** [review]

**Review by tgraf — APPROVED** [review]

File: pkg/bpf/map.go [hunk 25]
```diff
@@ -405,3 +395,17 @@ func (m *Map) DeleteAll() error {
...
 	return nil
 }
+
+//GetNextKey returns the next key in the Map after key.
+func (m *Map) GetNextKey(key MapKey, nextKey MapKey) error {
+	if err := m.Open(); err != nil {
+		return err
+	}
+
+	err := GetNextKey(m.fd, key.GetKeyPtr(), nextKey.GetKeyPtr())
```
Comment (aanm) [inline pkg/bpf/map.go]: can be `return GetNextKey(m.fd, key.GetKeyPtr(), nextKey.GetKeyPtr())`

File: pkg/maps/ctmap/ipv4.go [hunk 26]
```diff
@@ -0,0 +1,142 @@
...
+	flags   uint8
+}
+
+// GetKeyPtr returns the unsafe.Pointer for k.
+func (k *CtKey4) GetKeyPtr() unsafe.Pointer { return unsafe.Pointer(k) }
+
+//NewValue creates a new bpf.MapValue.
+func (k *CtKey4) NewValue() bpf.MapValue { return &CtEntry{} }
+
+// Convert converts CtKey4 fields between host bye order and map byte order.
```
Comment (aanm) [inline pkg/maps/ctmap/ipv4.go]: s/fields/ports

File: pkg/maps/ctmap/ipv4.go [hunk 27]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+// GetKeyPtr returns the unsafe.Pointer for k.
+func (k *CtKey4Global) GetKeyPtr() unsafe.Pointer { return unsafe.Pointer(k) }
+
+//NewValue creates a new bpf.MapValue.
+func (k *CtKey4Global) NewValue() bpf.MapValue { return &CtEntry{} }
+
+// Convert converts CtKey4Global fields between host bye order and map byte
+// order.
```
Comment (aanm) [inline pkg/maps/ctmap/ipv4.go]: s/fields/ports

File: pkg/maps/ctmap/ipv6.go [hunk 28]
```diff
@@ -0,0 +1,140 @@
...
+	flags   uint8
+}
+
+// GetKeyPtr returns the unsafe.Pointer for k.
+func (k *CtKey6) GetKeyPtr() unsafe.Pointer { return unsafe.Pointer(k) }
+
+//NewValue creates a new bpf.MapValue.
+func (k *CtKey6) NewValue() bpf.MapValue { return &CtEntry{} }
+
+// Convert converts CtKey6 fields between host bye order and map byte order.
```
Comment (aanm) [inline pkg/maps/ctmap/ipv6.go]: s/fields/ports

File: pkg/maps/ctmap/ipv6.go [hunk 29]
```diff
@@ -0,0 +1,140 @@
...
+	flags   uint8
+}
+
+// GetKeyPtr returns the unsafe.Pointer for k.
+func (k *CtKey6Global) GetKeyPtr() unsafe.Pointer { return unsafe.Pointer(k) }
+
+//NewValue creates a new bpf.MapValue.
+func (k *CtKey6Global) NewValue() bpf.MapValue { return &CtEntry{} }
+
+// Convert converts CtKey6Global fields between host bye order and map byte order.
```
Comment (aanm) [inline pkg/maps/ctmap/ipv6.go]: s/fields/ports

**Review by aanm — CHANGES_REQUESTED** [review]
Small nits, overall looks good

**Review by aanm — APPROVED** [review]

**Pull request merged**